teal/packer/plugin_types/git.tl

```diff
@@ -0,0 +1,464 @@
+    local lines = vim.split(commit_parts[1], '\n')
+    if is_breaking(body) or is_breaking(lines[2]) then
+      ret[#ret+1] = lines[1]
+    end
+  end
+  return ret
+end
+
+local function git_run(args: {string}, opts: jobs.Opts): jobs.JobResult
+  opts.env = opts.env or job_env
+  return jobs.run({ config.git.cmd, unpack(args) }, opts)
+end
+
+local function checkout(ref: string, opts: jobs.Opts, disp: Display): jobs.JobResult
+  if disp then
+    disp:task_update(fmt('checking out %s...', ref))
+  end
+  return git_run({'checkout', ref, '--' }, opts)
+end
+
+---@async
+local handle_checkouts = function(plugin: Plugin, disp: Display, opts: UpdateOpts): {string}
+  local function update_disp(msg: string)
+    if disp then
+      disp:task_update(plugin.full_name, msg)
+    end
+  end
+
+  update_disp('fetching reference...')
+
+  local job_opts: jobs.Opts = {
+    cwd = plugin.install_path
+  }
+
+  if plugin.tag and has_wildcard(plugin.tag) then
+    update_disp(fmt('getting tag for wildcard %s...', plugin.tag))
+    local jr = git_run({
+      'tag', '-l', plugin.tag,
+      '--sort', '-version:refname'
+    }, job_opts)
+    if jr:ok() then
+      local data = jr.output.data.stdout[1]
+      plugin.tag = vim.split(data, '\n')[1]
+    else
+      log.warn(fmt(
+        'Wildcard expansion did not found any tag for plugin %s: defaulting to latest commit...',
+        plugin.name
+      ))
+      plugin.tag = nil -- Wildcard is not found, then we bypass the tag
+      return jr.output.data.stderr
+    end
+  end
+
+  if (plugin.branch or (plugin.tag and not opts.preview_updates)) then
+    local branch_or_tag = plugin.branch or plugin.tag
+    local jr = checkout(branch_or_tag, job_opts, disp)
+    if not jr:ok() then
+      return jr.output.data.stderr
+    end
+  end
+
+  if plugin.commit then
+    local jr = checkout(plugin.commit, job_opts, disp)
+    if not jr:ok() then
+      return jr.output.data.stderr
+    end
+  end
+
+  return {}
+end
+
+local function split_messages(messages: {string}): {string}
+  local lines = {}
+  for _, message in ipairs(messages) do
+    vim.list_extend(lines, vim.split(message, '\n'))
+    table.insert(lines, '')
+  end
+  return lines
+end
+
+local function mark_breaking_changes(
+  plugin: Plugin,
+  disp: Display,
+  preview_updates: boolean
+): jobs.JobResult
+  disp:task_update(plugin.name, 'checking for breaking changes...')
+  local r = git_run({
+    'log',
+    '--color=never',
+    '--no-show-signature',
+    '--pretty=format:===COMMIT_START===%h%n%s===BODY_START===%b',
+    preview_updates and 'HEAD...FETCH_HEAD' or 'HEAD@{1}...HEAD'
+  }, {
+    cwd = plugin.install_path
+  })
+  if r:ok() then
+    plugin.breaking_commits = get_breaking_commits(r.output.data.stdout)
+  end
+  return r
+end
+
+local function get_install_cmd(plugin: Plugin): {string}
+  local install_cmd = {
+    'clone',
+    '--depth', tostring(plugin.commit and 999999 or config.git.depth),
+    '--no-single-branch',
```

Comment:
Question: is this option needed? It can inflate the checkout size, and I don't see why you would want to pull in all branches.

(If you switch to a branch in the spec, it's cleaner to remove and re-clone anyway. And `--depth 1` already prevents you from using the clone as an actual local repo.)

Reply:
Good question. No idea tbh. I'll add a todo for this.

Reply:
If it's a local plugin, no need to clone anyway ;)